### PR TITLE
Pulling updates from experimental

### PR DESCRIPTION
**Pull request by ajchili**

### CODE REVIEW

File: sync/src/com/kirinpatel/Main.java [hunk 1]
```diff
@@ -65,20 +66,18 @@ public Main() {
...
-        }
-
-        if (VersionChecker.isUpdated()) {
+        if (isUpdated()) {
             try {
                 UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
-            } catch (Exception e) {
-                Debug.Log("Unable to load \"Numbus\" UIManager!", 2);
+            } catch(Exception e) {
+                System.out.println("Unable to set look and feel of sync.");
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/Main.java]: UIMessage error?

File: sync/src/com/kirinpatel/Main.java [hunk 2]
```diff
@@ -187,9 +188,7 @@ public static void saveIPAddress(String ipAddress) {
     private static ArrayList<String> getPreviousAddresses() {
         File file = new File("launcherData.dat");
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/Main.java]: Use paths and the Files.newbufferedReader below

File: sync/src/com/kirinpatel/Main.java [hunk 3]
```diff
@@ -198,13 +197,62 @@ public static void saveIPAddress(String ipAddress) {
                 }
 
                 return ipAddresses;
-            }catch(IOException e) {
+            } catch (IOException e) {
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/Main.java]: UIMessage error

File: sync/src/com/kirinpatel/Main.java [hunk 4]
```diff
@@ -198,13 +197,62 @@ public static void saveIPAddress(String ipAddress) {
...
+            }
+
+            if (VERSION != v) return false;
+            else {
+                if (BUILD < b) return false;
+                else {
+                    if (REVISION < r && BUILD == b) return false;
+                }
+            }
+        } catch(MalformedURLException e) {
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/Main.java]: UIMessage error?

Comment (cltsd27) [inline sync/src/com/kirinpatel/Main.java]: Probably also return false here

File: sync/src/com/kirinpatel/Main.java [hunk 5]
```diff
@@ -198,13 +197,62 @@ public static void saveIPAddress(String ipAddress) {
...
+                        break;
+                    case 4:
+                        r = parsedInt;
+                        break;
+                    default:
+                        break;
+                }
+            }
+
+            if (VERSION != v) return false;
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/Main.java]: combine all of these using OR

File: sync/src/com/kirinpatel/gui/MediaSelectorGUI.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        setVisible(true);
+        isOpened = true;
+    }
+
+    /**
+     * Custom ActionListener that will serve to enable usability of MediaSelectorGUI JButtons.
+     */
+    class MediaButtonEvent implements ActionListener {
+
+        private int type;
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/gui/MediaSelectorGUI.java]: Enum instead

File: sync/src/com/kirinpatel/gui/MenuBar.java [hunk 7]
```diff
@@ -3,101 +3,75 @@
...
+              Share section
              */
             JMenuItem share = new JMenuItem("Share Server Address");
             share.addActionListener(e -> {
-                Toolkit.getDefaultToolkit().getSystemClipboard().setContents(new StringSelection(playbackPanel.type == 0 ? Server.ipAddress : Client.ipAddress), null);
-                new UIMessage("Server Address Copied", "The server address has been copied to your clipboard.", 0);
+                Toolkit.getDefaultToolkit()
+                        .getSystemClipboard()
+                        .setContents(
+                                new StringSelection(playbackPanel.type == 0
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/gui/MenuBar.java]: Enum for the type?

File: sync/src/com/kirinpatel/gui/PlaybackPanel.java [hunk 8]
```diff
@@ -196,18 +198,61 @@ public void mouseReleased(MouseEvent e) {
...
+
+        KeyDispatcher(PlaybackPanel playbackPanel) {
+            this.playbackPanel = playbackPanel;
+        }
+
+        @Override
+        public boolean dispatchKeyEvent(KeyEvent e) {
+            switch(e.getID()) {
+                case KeyEvent.KEY_PRESSED:
+                    switch(e.getKeyCode()) {
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/gui/PlaybackPanel.java]: Don't have to enum this, but put comments above the number indicating which key

File: sync/src/com/kirinpatel/gui/PlaybackPanel.java [hunk 9]
```diff
@@ -196,18 +198,61 @@ public void mouseReleased(MouseEvent e) {
...
+                            if (playbackPanel.isFullscreen && PlaybackPanel.pauseMedia.isEnabled()) {
+                                if (PlaybackPanel.mediaPlayer.getMedia().isPaused()) PlaybackPanel.mediaPlayer.play();
+                                else PlaybackPanel.mediaPlayer.pause();
+                            }
+                            break;
+                        case 122:
+                            if (playbackPanel.isFullscreen) playbackPanel.closeFullscreen();
+                            else  playbackPanel.initFullscreen();
+                            break;
+                        default:
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/gui/PlaybackPanel.java]: Are all these cases with just break necessary?

File: sync/src/com/kirinpatel/net/Media.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
+package com.kirinpatel.net;
+
+import java.io.Serializable;
+import java.nio.file.Path;
+
+public class Media implements Serializable {
+
+    private String url;
+    private String filePath;
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/net/Media.java]: Couldn't switch this over?

Comment (ajchili) [inline sync/src/com/kirinpatel/net/Media.java]: Was not possible as the User class implements a Media object. The connected users are sent from the server to clients and still cause the exception.

File: sync/src/com/kirinpatel/Main.java (hunk 1 above)
Comment (ajchili) [inline sync/src/com/kirinpatel/Main.java]: Fixed.

File: sync/src/com/kirinpatel/Main.java (hunk 4 above)
Comment (ajchili) [inline sync/src/com/kirinpatel/Main.java]: Fixed.

File: sync/src/com/kirinpatel/gui/MenuBar.java (hunk 7 above)
Comment (ajchili) [inline sync/src/com/kirinpatel/gui/MenuBar.java]: Fixed.

File: sync/src/com/kirinpatel/gui/PlaybackPanel.java (hunk 9 above)
Comment (ajchili) [inline sync/src/com/kirinpatel/gui/PlaybackPanel.java]: Fixed.

File: sync/src/com/kirinpatel/gui/PlaybackPanel.java (hunk 8 above)
Comment (ajchili) [inline sync/src/com/kirinpatel/gui/PlaybackPanel.java]: Did not implement enums.

File: sync/src/com/kirinpatel/gui/MediaSelectorGUI.java (hunk 6 above)
Comment (ajchili) [inline sync/src/com/kirinpatel/gui/MediaSelectorGUI.java]: Fixed.

File: sync/src/com/kirinpatel/Main.java (hunk 5 above)
Comment (ajchili) [inline sync/src/com/kirinpatel/Main.java]: Fixed.

File: sync/src/com/kirinpatel/Main.java (hunk 2 above)
Comment (ajchili) [inline sync/src/com/kirinpatel/Main.java]: Fixed.

File: sync/src/com/kirinpatel/gui/PlaybackPanel.java [hunk 11]
```diff
@@ -227,27 +244,34 @@ public boolean dispatchKeyEvent(KeyEvent e) {
...
+                            if (playbackPanel.isFullscreen) {
+                                playbackPanel.closeFullscreen();
+                            } else  {
+                                playbackPanel.initFullscreen();
+                            }
                             break;
                         default:
                             break;
                     }
                     break;
```
Comment (cltsd27) [inline sync/src/com/kirinpatel/gui/PlaybackPanel.java]: Is this break doing anything

Comment (ajchili) [inline sync/src/com/kirinpatel/gui/PlaybackPanel.java]: The break is required.

**Comment by ajchili** [comment]
Playback panel has been moved to a singleton instance, however, a new implementation of server/client restarts are required. Currently, with the new push, the sync application is closed and needs to be restarted on server stop or client disconnect. The PlaybackPanel, as a singleton instance, requires that a new VLCMediaPlayer is created on the creation of GUI to prevent issues such as `NullPointerException` and `Invalid memory access`.

**Comment by ajchili** [comment]
The PlabackPanel class is no longer truly a singleton instance as it would be too much work to implement a solution with the PlaybackPanel as a singleton instance. Look [here](https://github.com/ajchili/sync/pull/26/commits/7a6054f486a8ba298f43224c28a102d32ea6b1df#diff-20c41bd92e661bd77a0edeb355456a0fR45) for reference.

**Pull request merged**